omniledger/service/contracts.go

```diff
@@ -19,6 +20,9 @@ var ZeroDarc = darc.ID(make([]byte, 32))
 // ConfigID is 64 bytes of zeroes.
 var ConfigID = ObjectID{ZeroDarc, ZeroNonce}
 
+// BlockIntervalID is the key to the block interval information.
+var BlockIntervalID = ObjectID{darc.ID([]byte("block_interval")), ZeroNonce}
```

Comment:
I would put the block interval into the `ConfigID` object/account. We will have other things that we want to configure, and if we create another object for each configuration, there might be too many objects.

So the `ConfigID` object has a value that is a marshalled `struct{ GenesisDarc darc.Darc, BlockInterval time.Duration}`. And we need to find a good way to describe the data-structures stored in the contracts to access them from the outside.

Reply:
The reason I created a new ID is for that exact reason. We will need to unmarshal all of the config to just read one int64. I can try to keep it in memory upon creation.

Reply:
OK, I see. I was worried about marshalling/unmarshalling for events that come in at 100s per second. For the configuration, it will be much less, so I think it's more convenient to have a `Configuration` structure that is marshalled/unmarshalled as needed.

When we discussed the ConfigID first, we only thought about the genesis-darc, but now other data will also have to be stored. One thing we could do is the following (plus renaming the variables ;):

## 1 - One Object for all configuration

`var GenesisDarcIDObjectID = ObjectID{ZeroDarc, ZeroNonce}` holds the ID of the genesis-darc
`var GenesisDarcObjectID = ObjectID{GenesisDarcID, ZeroNonce}` holds the genesis-darc itself so that it can be evolved
`var ConfigObjectID = ObjectID{GenesisDarcID, OneNonce}` holds the configuration (BlockInterval, Roster(?), IdentitySkipchain, CoinsMapping (CPU, Memory), ...)

So if you want to look up the configuration, you have to go through the `GenesisDarcIDObjectID` (bad name, but it explains nicely what it is), to get the `ConfigObjectID` and then you can read the configuration.

## 2 - One object for every configuration parameter

The alternative you propose is to have
`var ConfigIntervalObjectID = ObjectID{GenesisDarc, OneNonce}`
`var ConfigRosterObjectID = ObjectID{GenesisDarc, TwoNonce}`
...
and then each `Config.*ObjectID` will have to have its own contract that defines how you're allowed to change that object.

## Conclusion

I prefer 1) because it needs less contracts and gives a 'central' configuration that you can change. Using the actions in the genesis-darc you can define who is allowed to change which configuration parameter, as each parameter will have its own command (`UpdateInterval`, `UpdateRoster`, ...)